### ISSUE TITLE
训练矩形灰度图时出错

### ISSUE THREAD

**Issue by sky-2016yty** [PaddlePaddle/Paddle#10523]
训练的图片是640*180灰度图,我通过paddle的API(paddle.image.center_crop())从中间截取出128*128的图片(之前随机截取会报数据与数据层不一致),开始训练后报错:
F0509 17:48:53.358350 17111 Matrix.cpp:3688] Check failed: (size_t)lbl[i] < dim (5 vs. 5) 
*** Check failure stack trace: ***
    @     0x7efc443329cd  google::LogMessage::Fail()
    @     0x7efc4433647c  google::LogMessage::SendToLog()
    @     0x7efc443324f3  google::LogMessage::Flush()
    @     0x7efc4433798e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7efc44282422  paddle::CpuMatrix::oneHotCrossEntropy()
    @     0x7efc440a8e15  paddle::CostLayer::forward()
    @     0x7efc440e7efd  paddle::NeuralNetwork::forward()
    @     0x7efc441096e4  paddle::TrainerThread::forward()
    @     0x7efc4410a945  paddle::TrainerThread::computeThread()
    @     0x7efc56ec4ae4  (unknown)
    @     0x7efc6fcd36db  start_thread
    @     0x7efc7000c88f  clone
    @              (nil)  (unknown)
这个问题是因为图片的尺寸原因还是灰度图的原因?

**Comment by qingqing01** [comment]
从错误信息来看，是label的index超范围了，注意Paddle里的label的index是从0开始的。

**Comment by sky-2016yty** [comment]
谢谢 问题解决了 ,但是我如果用矩形的图片,比如用64*14的图片就报错
I0509 19:28:27.223096 23062 Util.cpp:166] commandline:  --use_gpu=False --trainer_count=4 
[INFO 2018-05-09 19:28:27,228 layers.py:2689] output for __conv_0__: c = 64, h = 18, w = 64, size = 73728
[INFO 2018-05-09 19:28:27,228 layers.py:2689] output for __conv_1__: c = 64, h = 18, w = 64, size = 73728
[INFO 2018-05-09 19:28:27,229 layers.py:2829] output for __pool_0__: c = 64, h = 9, w = 32, size = 18432
[CRITICAL 2018-05-09 19:28:27,230 layers.py:2689] Input layer __pool_0__: Incorrect input image size 32 * 9 for input image pixels 18432
网络模型为:
def vgg_bn_drop(datadim, type_size,shape):
    # 输入数据模式
    image = paddle.layer.data(name="image",
                              type=paddle.data_type.dense_vector(datadim),
                              height=18,
                              width=64)

    def conv_block(ipt, num_filter, groups, dropouts, num_channels=1):
        return paddle.networks.img_conv_group(
            input=ipt,
            num_channels=num_channels,
            pool_size=2,
            pool_stride=2,
            conv_num_filter=[num_filter] * groups,
            conv_filter_size=3,
            conv_act=paddle.activation.Relu(),
            conv_with_batchnorm=False,
            conv_batchnorm_drop_rate=dropouts,
            pool_type=paddle.pooling.Max())

    conv1 = conv_block(image, 64, 2, [0.3, 0], 1)
    conv2 = conv_block(conv1, 128, 2, [0.4, 0])
    conv3 = conv_block(conv2, 256, 3, [0.4, 0.4, 0])
    conv4 = conv_block(conv3, 512, 3, [0.4, 0.4, 0])
    conv5 = conv_block(conv4, 512, 3, [0.4, 0.4, 0])

    drop = paddle.layer.dropout(input=conv5, dropout_rate=0.5)
    fc1 = paddle.layer.fc(input=drop, size=512, act=paddle.activation.Linear())
    bn = paddle.layer.batch_norm(input=fc1,
                                 act=paddle.activation.Relu(),
                                 layer_attr=paddle.attr.Extra(drop_rate=0.5))
    fc2 = paddle.layer.fc(input=bn, size=512, act=paddle.activation.Linear())
   
    out = paddle.layer.fc(input=fc2,
                          size=type_size,
                          act=paddle.activation.Softmax())
    return out

**Comment by qingqing01** [comment]
> def conv_block(ipt, num_filter, groups, dropouts, num_channels=1):

这里的`num_channels=1`需要换成`num_channels=None`，后面的conv层，会依据输入层推断出输入channel数是多少。

否则，对于 `conv2 = conv_block(conv1, 128, 2, [0.4, 0])` 这里conv也会设置输入的channel是1，其实是64。

其实从错误信息里：

> [INFO 2018-05-09 19:28:27,229 layers.py:2829] output for pool_0: c = 64, h = 9, w = 32, size = 18432
> [CRITICAL 2018-05-09 19:28:27,230 layers.py:2689] Input layer pool_0: Incorrect input image size 32 * 9 for input image pixels 18432
也能推断出32 * 9少乘了channel 64，是某层的channel数设置错误。

下面是我正确的测试程序，改成None即可：

```python
import sys
import paddle.v2 as paddle
import numpy as np

def vgg_bn_drop(image, type_size):
    def conv_block(ipt, num_filter, groups, dropouts, num_channels=None):
        return paddle.networks.img_conv_group(
            input=ipt,
            num_channels=num_channels,
            pool_size=2,
            pool_stride=2,
            conv_num_filter=[num_filter] * groups,
            conv_filter_size=3,
            conv_act=paddle.activation.Relu(),
            conv_with_batchnorm=False,
            conv_batchnorm_drop_rate=dropouts,
            pool_type=paddle.pooling.Max())

    conv1 = conv_block(image, 64, 2, [0.3, 0], 1)
    conv2 = conv_block(conv1, 128, 2, [0.4, 0])
    conv3 = conv_block(conv2, 256, 3, [0.4, 0.4, 0])
    conv4 = conv_block(conv3, 512, 3, [0.4, 0.4, 0])
    conv5 = conv_block(conv4, 512, 3, [0.4, 0.4, 0])

    drop = paddle.layer.dropout(input=conv5, dropout_rate=0.5)
    fc1 = paddle.layer.fc(input=drop, size=512, act=paddle.activation.Linear())
    bn = paddle.layer.batch_norm(input=fc1,
                                 act=paddle.activation.Relu(),
                                 layer_attr=paddle.attr.Extra(drop_rate=0.5))
    fc2 = paddle.layer.fc(input=bn, size=512, act=paddle.activation.Linear())

    out = paddle.layer.fc(input=fc2,
                          size=type_size,
                          act=paddle.activation.Softmax())
    return out

def train_reader():
    def reader():
        data = np.random.normal(size=(18 * 64)).astype('float32').flatten()
        yield data, 1
    return reader


def main():
    datadim = 1 * 18 * 64
    classdim = 10

    # PaddlePaddle init
    paddle.init(use_gpu=False, trainer_count=1)
    image = paddle.layer.data(name="image",
                              type=paddle.data_type.dense_vector(datadim),
                              height=18,
                              width=64)
    out = vgg_bn_drop(image, classdim)
    lbl = paddle.layer.data(
        name="label", type=paddle.data_type.integer_value(classdim))
    cost = paddle.layer.classification_cost(input=out, label=lbl)

    # Create parameters
    parameters = paddle.parameters.create(cost)

    # Create optimizer
    momentum_optimizer = paddle.optimizer.Momentum(
        momentum=0.9,
        regularization=paddle.optimizer.L2Regularization(rate=0.0002 * 128),
        learning_rate=0.1 / 128.0,
        learning_rate_decay_a=0.1,
        learning_rate_decay_b=50000 * 100,
        learning_rate_schedule='discexp')

    # End batch and end pass event handler
    def event_handler(event):
        if isinstance(event, paddle.event.EndIteration):
            print "\nPass %d, Batch %d, Cost %f, %s" % (
                event.pass_id, event.batch_id, event.cost, event.metrics)

    # Create trainer
    trainer = paddle.trainer.SGD(
        cost=cost, parameters=parameters, update_equation=momentum_optimizer)
    trainer.train(reader=paddle.batch(train_reader(), batch_size=32),
                  num_passes=200,
                  event_handler=event_handler,
                  feeding={'image': 0,
                           'label': 1})


if __name__ == '__main__':
    main()

```

**Comment by sky-2016yty** [comment]
十分感谢大神
